### DISCUSSION TITLE
Config for dark mode

### DISCUSSION THREAD

**Discussion by inorganik** [tailwindlabs/tailwindcss#10408] [Ideas]
Reading the [dark mode docs](https://tailwindcss.com/docs/dark-mode) I was surprised to see that you can't simply set some variables for color values and let them get applied to all the styles to support dark mode. 

Having to apply classes to elements on a one-off basis sounds ridiculous, especially when you can do it in one file in other CSS frameworks like Bootstrap and Ionic. Here's what it looks like in [ionic](https://github.com/ionic-team/ionic-conference-app/blob/master/src/theme/variables.scss#L276). Could some variables be exposed in the config under a `darkMode` key, which would modify the colors used on various classes without requiring the `dark:` prefix?

Having this config would provide the following benefits:

- improve maintainability
- greatly reduce markup clutter
- improve developer experience

**Comment by adamwathan** [comment]
Hey! Check out this thread I posted on Twitter the other day, it explains why dark mode is designed the way it is in Tailwind and how to configure it to use variables the way you would like:

https://twitter.com/adamwathan/status/1614811345941061634?s=46&t=-_D8LZ-llqUwzAc75iugVA

**Comment by inorganik** [comment]
Thank you for the reply, I read your thread. Defining your colors as css variables seems like a good solution, but how do I know which colors to define?

In your thread you gave an example of github setting some 400 vars, I think that is a bit of an overblown example. If you look at my ionic example, it's around ~50-60 values, which are just variations of 9 colors. You mention flexibility and control, but that flexibility should be achieved by case-by-case overrides, rather than requiring _every single element_ to be hand-styled.

In other words, you still achieve the same flexibility but you only need to apply special `dark:` classes where an override is needed, instead of doing it every where.

**Reply by adamwathan** [reply]
> Defining your colors as css variables seems like a good solution, but how do I know which colors to define?

That's kinda the whole thing, this is totally dependent on your app and what contexts you need colors for. If your UI has cards maybe you need a `--card-bg` variable, if it has badges you probably need a `--badge-bg` variable, if the badges have statuses specific to the domain of your app you might need `--badge-payment-due` and `--badge-paid-on-time` for example.

> In other words, you still achieve the same flexibility but you only need to apply special dark: classes where an override is needed, instead of doing it every where.

This is exactly what I was showing how to do in this tweet:

https://twitter.com/adamwathan/status/1614811352861929472

Here's a Tailwind Play with a simple real-world example:

https://play.tailwindcss.com/o0QnVG2qG0

You can see the entire thing is powered by CSS variables and there is no use of the `dark:` modifier, but all of the variables are very specific to this UI, which is why there aren't really any variables we could ship by default that are going to work for everyone.

You'll also see the mappings are not always just simply inverting things. For instance, card background color goes from `white` to `gray-800`, but the secondary button background color goes from `white` to `gray-600`, so it's not as simple as just "white should always flip to gray-800 no matter what it's being used for" — the changes are always UI-specific like I mentioned in the tweet.

**Comment by inorganik** [comment]
I agree you should never just invert colors. But I think you're providing great examples of some default values to ship. If you were forced under some strange circumstance to ship some default variables, could you do it? Actually this would be a great feature for TW 4.0. Picture it, "Tailwind dramatically speeds up your frontend work. Now it's even faster with _tailwind keys_. Unlock more power by setting new config variables for 1 or more themes" 😄

You could ship with 3 background shades, 3 text shades, an outline color, and then theme color variables. There are projects like [tailwindcss-themer](https://github.com/RyanClementsHax/tailwindcss-themer) that have done this and seem to be pretty popular. But perhaps your intention is to keep it more basic, more of a building block? 

Either way I appreciate the time you've taken to discuss it here. I think it benefits everyone to have this discussion because theming is much more difficult in TW than other css frameworks and people probably wonder how we could make it better.

**Reply by adamwathan** [reply]
I think that sort of thing is better served by the plugin ecosystem, as Tailwind itself was designed from the beginning to be useful for building _any_ website and not to be prescriptive. The goal is to take literally any design file someone throws over the wall to you and build it faster than you could with raw CSS.

>  theming is much more difficult in TW than other css frameworks

I don't really think this is true, unless you are totally content with letting the framework you're using drive how your project looks. If you design a light mode design and a dark mode design with no constraints in a design tool, no other CSS framework will or can make that easier to implement than Tailwind, because there's simply nothing about the design you can predict.

Personally I never want to make Tailwind worse for building custom designs for any reason — the second it becomes opinionated about your actual design it's not an interesting tool IMO.

**Reply by inorganik** [reply]
Gotcha. I personally feel there's room to set some base values, but this is your baby. Thanks for the discussion!

If anyone ends up here looking for theming resources, I found some good ones:

- [Daisy UI](https://daisyui.com/)
- [tailwindcss themer](https://github.com/RyanClementsHax/tailwindcss-themer)
- [Nightwind](https://nightwindcss.com/)
